src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
                     }
-                    else
+                    else if (controls[i].Type == "1.2.840.113556.1.4.1504")
                     {
-                        // decoding might fail as attribute is optional
-                        o = BerConverter.Decode("{e}", value);
-                        Debug.Assert(o != null && o.Length == 1);
+                        // The control is an AsqControl, as described in MS-ADTS section 3.1.1.3.4.1.18
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                        ThrowUnless(sequenceContentLength > 0);
 
-                        result = (int)o[0];
-                    }
+                        ResultCode result = AsnDecoder.ReadEnumeratedValue<ResultCode>(asnSpan.Slice(sequenceContentOffset, sequenceContentLength), AsnEncodingRules.BER, out _);
 
-                    SortResponseControl sort = new SortResponseControl((ResultCode)result, attribute, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = sort;
-                }
-                else if (controls[i].Type == "2.16.840.1.113730.3.4.10")
-                {
-                    // The control is a VlvResponseControl.
-                    int position;
-                    int count;
-                    int result;
-                    byte[] context = null;
-                    object[] o = BerConverter.TryDecode("{iieO}", value, out bool decodeSucceeded);
-
-                    if (decodeSucceeded)
+                        AsqResponseControl asq = new AsqResponseControl(result, controls[i].IsCritical, value);
+                        controls[i] = asq;
+                    }
+                    else if (controls[i].Type == "1.2.840.113556.1.4.841")
                     {
-                        Debug.Assert(o != null && o.Length == 4);
-                        position = (int)o[0];
-                        count = (int)o[1];
-                        result = (int)o[2];
-                        context = (byte[])o[3];
+                        // The control is a DirSyncControl, as described in MS-ADTS section 3.1.1.3.4.1.3
+                        byte[] dirsyncCookie;
+
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                        ThrowUnless(sequenceContentLength > 0);
+                        asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
+
+                        asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int moreData, out int bytesConsumed);
+                        ThrowUnless(asnReadSuccessful);
+                        asnSpan = asnSpan.Slice(bytesConsumed);
+
+                        asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int count, out bytesConsumed);
```

Comment:
The referenced spec calls this `MaxBytes`, and says that any value less than `0x100000` should be treated as `0x100000`.  I don't see that minimum-inflation here, nor in the constructor, so I assume it's a caller-interpretation problem.

This isn't changing any code flows around, so nothing _really_ needs to be done about it, but since I read it while wondering if Int32 was a valid assumption I dug a little deeper and was surprised.

Reply:
Since we're referring to a `DirSyncResponseValue`, the actual BER field name is `unused`. This doesn't have the same minimum-inflation logic behind it, but does raise the question of why exactly we read a field which the spec defines as being unused and treat it as a count. I don't think there's any action to be taken, but it's a little odd.